### PR TITLE
Added accumulation of loggers' metrics for the same steps

### PR DESCRIPTION
**Pull request by alexeykarnachev**
# Before submitting

- [x] Was this discussed/approved via a Github issue? [#1173](https://github.com/PyTorchLightning/pytorch-lightning/issues/1173)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CHANGELOG.md)?

## What does this PR do?
Fixes [#1173](https://github.com/PyTorchLightning/pytorch-lightning/issues/1173).
This PR adds new method `LightningLoggerBase.agg_and_log_metrics`. This method has the same signature as and abstract `LightningLoggerBase.log_metrics` method.
New method aggregates metrics with the same `step` value and then, delegates the logging procedure to `log_metrics`.
From the client perspective nothing has changed. An abstract `log_metrics` still in place and it still needs to be implemented in the child loggers, but the `Trainer` now calls the `agg_and_log_metrics`.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by alexeykarnachev** [comment]
Oh, I did the merge instead of rebase. sry

File: pytorch_lightning/loggers/base.py [hunk 1]
```diff
@@ -27,20 +31,62 @@ class LightningLoggerBase(ABC):
...
+            Aggregated metrics. The return value could be None. In such case, metrics
+            are added to the aggregation list, but not aggregated yet.
+        """
+        if step == self._prev_step:
+            self._metrics_to_agg.append(collections.Counter(metrics))
+            return None
+
+        if len(self._metrics_to_agg) > 0:
+            agg_mets = functools.reduce(lambda c1, c2: c1 + c2, self._metrics_to_agg)
+            agg_mets = {k: v / len(self._metrics_to_agg) for k, v in agg_mets.items()}
```
Comment (jeremyjordan) [inline pytorch_lightning/loggers/base.py]: should we factor out the aggregation function as an op (similar to [here[(https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/callbacks/early_stopping.py#L59))? and perhaps the reduction op is passed into the logger at initialization? i think a `mean` reduction will be most common, but it might be nice to be able to do `min` and `max` reductions as well.

Comment (alexeykarnachev) [reply pytorch_lightning/loggers/base.py]: good idea! done!

File: pytorch_lightning/loggers/metrics_agg.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
+import functools
+import operator
+from typing import Dict, Callable, Mapping, Sequence
+
+MetricsT = Dict[str, float]
```
Comment (Borda) [inline pytorch_lightning/loggers/metrics_agg.py]: do we really need this new type?

File: pytorch_lightning/loggers/metrics_agg.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        >>> d1 = {'a': 1.7, 'b': 2.0, 'c': 1}
+        >>> d2 = {'a': 1.1, 'b': 2.2}
+        >>> fn = max
+        >>> pprint.pprint(merge_two_dicts(d1, d2, fn))
+        {'a': 1.7, 'b': 2.2, 'c': 1}
+    """
+
+    all_keys = list(set(d1.keys()).union(set(d2.keys())))
+
+    return {key: fn(d1.get(key, default_value), d2.get(key, default_value)) for key in all_keys}
```
Comment (Borda) [inline pytorch_lightning/loggers/metrics_agg.py]:
```suggestion
    keys = set(list(d1.keys()) + list(d2.keys()))
    dx = {k: fn([v for v in (d1.get(k), d2.get(k)) if v]) for k in keys}
    return dx
```

File: pytorch_lightning/loggers/base.py [hunk 4]
```diff
@@ -25,22 +27,64 @@ def wrapped_fn(self, *args, **kwargs):
...
     @property
     @abstractmethod
     def experiment(self) -> Any:
         """Return the experiment object associated with this logger"""
 
+    def _agg_metrics(self, metrics: metrics_agg.MetricsT, step: Optional[int] = None) -> Optional[metrics_agg.MetricsT]:
+        """Aggregates metrics.
+
+        Args:
+             Check `log_metrics` method for the arguments description.
```
Comment (Borda) [inline pytorch_lightning/loggers/base.py]: you should list the arguments explicitly

File: pytorch_lightning/loggers/base.py [hunk 5]
```diff
@@ -25,22 +27,64 @@ def wrapped_fn(self, *args, **kwargs):
...
+        self._prev_step = step
+        return agg_mets
+
+    def agg_and_log_metrics(self, metrics: metrics_agg.MetricsT, step: Optional[int] = None):
+        """Aggregates and records metrics.
+        This method doesn't log the passed metrics instantaneously, but instead
+        it aggregates them and logs only if metrics are ready to be logged.
+
+        Args:
+             Check `log_metrics` method for the arguments description.
```
Comment (Borda) [inline pytorch_lightning/loggers/base.py]: you should list the arguments explicitly

File: pytorch_lightning/loggers/metrics_agg.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
+import functools
+import operator
+from typing import Dict, Callable, Mapping, Sequence
+
+MetricsT = Dict[str, float]
+MetricsAggFnT = Callable[[Sequence[MetricsT]], MetricsT]
+
+
+def merge_two_dicts(d1: Mapping, d2: Mapping, fn: Callable[[float, float], float], default_value: float = 0) -> Dict:
+    """Merges two dictionaries values with the given function.
```
Comment (Borda) [inline pytorch_lightning/loggers/metrics_agg.py]: what is different metrics are aggregated differently max/mean

File: tests/loggers/test_metrics_agg.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
+import operator
```
Comment (Borda) [inline tests/loggers/test_metrics_agg.py]: not sure if we need these test since it already tested via doctest

File: pytorch_lightning/loggers/metrics_agg.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexeykarnachev) [reply pytorch_lightning/loggers/metrics_agg.py]: I refactored it a little bit

File: pytorch_lightning/loggers/metrics_agg.py (hunk 2 above)
Comment (alexeykarnachev) [reply pytorch_lightning/loggers/metrics_agg.py]: removed

File: pytorch_lightning/loggers/metrics_agg.py (hunk 6 above)
Comment (alexeykarnachev) [reply pytorch_lightning/loggers/metrics_agg.py]: As we discussed, the main, or at least the one which I can see purpose of such the aggregation is to log accumulation batches correctly. In such case, the most (singe?) natural way of aggregation is the averaging across batches (metrics for identical steps). So, it seems like one aggregation strategy might be enough.
If you see another options (not only the averaging of accum. steps), we can add dict of functions (one function for each metric name). And if it's None, use the default strategy for all metrics. What do you think?

File: tests/loggers/test_metrics_agg.py (hunk 7 above)
Comment (alexeykarnachev) [reply tests/loggers/test_metrics_agg.py]: Yes, you are right. All stuff is tested in docstrings. I removed the tests file.

File: pytorch_lightning/loggers/metrics_agg.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    Examples:
+        >>> import pprint
+        >>> metrics_to_agg = [{'a': 1.7, 'b': 2.0}, {'a': 1.1, 'b': 2.2}, {'a': 0.0, 'b': 0.1, 'c': 0.2}]
+        >>> pprint.pprint(metrics_agg_sum(metrics_to_agg))
+        {'a': 2.8, 'b': 4.3, 'c': 0.2}
+    """
+    return merge_dicts(metrics_to_agg, sum)
+
+
+def metrics_agg_max(metrics_to_agg: Sequence[Dict[str, float]]) -> Dict[str, float]:
```
Comment (Borda) [inline pytorch_lightning/loggers/metrics_agg.py]: not sure if we need this wrapping just for one changed parameters, I would rather keep it passing as parameter

File: pytorch_lightning/loggers/base.py [hunk 9]
```diff
@@ -25,22 +27,64 @@ def wrapped_fn(self, *args, **kwargs):
 class LightningLoggerBase(ABC):
     """Base class for experiment loggers."""
 
-    def __init__(self):
+    def __init__(self, metrics_agg_fn: Callable[[Sequence[Dict[str, float]]], float] = metrics_agg.metrics_agg_avg):
```
Comment (Borda) [inline pytorch_lightning/loggers/base.py]: this means that a single logger can do only one type of operation (max/avg/..) overall metrics?

File: pytorch_lightning/loggers/metrics_agg.py (hunk 8 above)
Comment (jeremyjordan) [reply pytorch_lightning/loggers/metrics_agg.py]: agreed, i don't think we need a separate function for each `reduce_fn` option.

File: pytorch_lightning/loggers/base.py (hunk 9 above)
Comment (justusschock) [reply pytorch_lightning/loggers/base.py]: I agree, we should have at least the option to give a separate aggregation function per key (as a dict) and when only one function is given, we can still build an internal dict so that we always do a lookup there.

File: pytorch_lightning/loggers/metrics_agg.py (hunk 6 above)
Comment (Borda) [reply pytorch_lightning/loggers/metrics_agg.py]: @justusschock pls ^^

**Comment by Borda** [comment]
@justusschock I think that it would be nice to help @alexeykarnachev with growing some aggregation which may be later useful also for metrics than shaking there and back... do you have any suggestion/idea?

**Comment by justusschock** [comment]
@Borda @alexeykarnachev back when I've written my own framework I solved it like this (of course replace the numpy function with something appropriate/torch functions):
https://github.com/delira-dev/delira/blob/master/delira/utils/dict_reductions.py

Then each logger was given a dict of key, reduce_specifier like this:
```
{'key1': 'max', 'key2': 'min'} and we had a fallback function (mean) for non-specified keys.
We basically did this, because you would probably want to log scalars more often then images. And we used the same for metric reduction.

For reference, this was our whole logger implementation (in some parts not the most beautiful code :D ):
https://github.com/delira-dev/delira/blob/master/delira/logging/base_logger.py

File: pytorch_lightning/loggers/metrics_agg.py (hunk 6 above)
Comment (justusschock) [reply pytorch_lightning/loggers/metrics_agg.py]: answered below, since this is not just an implementation detail, but a more fundamental decision on logging

**Comment by williamFalcon** [comment]
@Borda @alexeykarnachev @justusschock can we get this merged soon so we can add to 0.7.2?

**Comment by Borda** [comment]
not yet we have some API disagreement...

**Comment by alexeykarnachev** [comment]
planing to add all discussed fixes tomorrow

**Comment by alexeykarnachev** [comment]
Now, there are two arguments in the `LightningLoggerBase`. They have default values, so the 
inherited classes won't break when they call `super`.
But what do you think about adding these arguments to all inherited classes constructors explicitly (also with default value)?

File: pytorch_lightning/loggers/base.py [hunk 10]
```diff
@@ -22,25 +25,136 @@ def wrapped_fn(self, *args, **kwargs):
...
+            default function).
+        dflt_func (function):
+            Default function to aggregate keys, which are not presented in the
+            `agg_funcs` map.
+
+    Returns (dict):
+        Dictionary with merged values.
+
+    Examples:
+        >>> import pprint
```
Comment (awaelchli) [inline pytorch_lightning/loggers/base.py:53]: doctest, NICE!!

File: CHANGELOG.md [hunk 11]
```diff
@@ -21,6 +22,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Added `summary` method to Profilers. ([#1259](https://github.com/PyTorchLightning/pytorch-lightning/pull/1259))
 - Added informative errors if user defined dataloader has zero length ([#1280](https://github.com/PyTorchLightning/pytorch-lightning/pull/1280))
 - Added testing for python 3.8 ([#915](https://github.com/PyTorchLightning/pytorch-lightning/pull/915)) 
+- Added support for `IterableDataset` when `val_check_interval=1.0` (default), this will trigger validation at the end of each epoch. ([#1283](https://github.com/PyTorchLightning/pytorch-lightning/pull/1283))
```
Comment (awaelchli) [inline CHANGELOG.md:25]: if you do a rebase, these unrelated changes will disappear :)

Comment (alexeykarnachev) [reply CHANGELOG.md:25]: Now I did the rabase, but there are a lot of commits have appeared (see below). I think, I need to go to read rebase tutorial :)

Comment (awaelchli) [reply CHANGELOG.md:25]: Hmm, I didn't mean to make you more work, sorry :) I don't know what happened, but the good thing is you can always checkout the commit before you did the rebase, and then force push. This will revert it again to what you had before.

**Comment by Borda** [comment]
@alexeykarnachev thx for your work, could you do proper rebase, not it seems that you only merged master as we can see about 30 commits from master yielding in 50 changed files...

File: CHANGELOG.md (hunk 11 above)
Comment (Borda) [reply CHANGELOG.md:25]: https://github.com/PyTorchLightning/pytorch-lightning/pull/1278#issuecomment-609016922

**Comment by alexeykarnachev** [comment]
Yes, sry) I'll handle this

**Comment by Borda** [comment]
> Yes, sry) I'll handle this

That's fine, just helping each other lol

**Comment by alexeykarnachev** [comment]
@Borda could you check, please, is it okay now?

**Comment by Borda** [comment]
> @Borda could you check, please, is it okay now?

Cool, not is it correct, Thx! I ll check the content now :]

File: pytorch_lightning/trainer/supporting_classes.py [hunk 12]
```diff
@@ -0,0 +1,40 @@
+import torch
```
Comment (Borda) [inline pytorch_lightning/trainer/supporting_classes.py:1]: this class already exists

File: pytorch_lightning/loggers/base.py [hunk 13]
```diff
@@ -22,25 +25,136 @@ def wrapped_fn(self, *args, **kwargs):
...
+
+
 class LightningLoggerBase(ABC):
     """Base class for experiment loggers."""
 
-    def __init__(self):
+    def __init__(
+            self,
+            agg_funcs: Optional[Mapping[str, Callable[[Sequence[float]], float]]] = None,
+            dflt_func: Callable[[Sequence[float]], float] = np.mean
```
Comment (Borda) [inline pytorch_lightning/loggers/base.py:79]: this shall be depending on metric, not one for all in single logger...

File: pytorch_lightning/loggers/base.py [hunk 14]
```diff
@@ -22,25 +25,136 @@ def wrapped_fn(self, *args, **kwargs):
...
+            agg_mets = merge_dicts(
+                self._metrics_to_agg, self._agg_funcs, self._dflt_func)
+        else:
+            agg_mets = None
+
+        self._metrics_to_agg = [metrics]
+        self._prev_step = step
+        return agg_mets
+
+    def agg_and_log_metrics(self, metrics: Dict[str, float], step: Optional[int] = None):
```
Comment (Borda) [inline pytorch_lightning/loggers/base.py:133]: like we can specify the mean/avg/... here

File: pytorch_lightning/loggers/base.py (hunk 13 above)
Comment (alexeykarnachev) [reply pytorch_lightning/loggers/base.py:79]: Could you, clarify a little? Did not get your idea

**Comment by williamFalcon** [comment]
@borda let's get this PR merged ASAP

File: pytorch_lightning/trainer/trainer.py [hunk 15]
```diff
@@ -551,20 +552,20 @@ def add_argparse_args(cls, parent_parser: ArgumentParser) -> ArgumentParser:
...
-                            default=arg_default,
-                            type=allowed_type,
-                            dest=arg,
-                            help='autogenerated by pl.Trainer'
-                        )
-                        break
+        for arg, arg_types, arg_default in (at for at in cls.get_init_arguments_and_types()
+                                            if at[0] not in depr_arg_names):
+            for allowed_type in (at for at in allowed_types
+                                 if allowed_type in arg_types):
```
Comment (alexeykarnachev) [inline pytorch_lightning/trainer/trainer.py:558]: There is not `allowed_type` variable in the scope of this generator: 
`(at for at in allowed_types if allowed_type in arg_types)`
Maybe it should be like this:
`(at for at in allowed_types if at in arg_types)` ?

**Comment by williamFalcon** [comment]
@alexeykarnachev  we're preparing the 0.7.2 release for tomorrow. would be good to get this in there :)

File: pytorch_lightning/trainer/trainer.py (hunk 15 above)
Comment (Borda) [reply pytorch_lightning/trainer/trainer.py]: true, my bad

**Comment by Borda** [comment]
@alexeykarnachev GREAT job and sorry for overwriting some of your work...

File: pytorch_lightning/loggers/base.py [hunk 16]
```diff
@@ -200,3 +303,48 @@ def name(self) -> str:
...
+    Returns:
+        Dictionary with merged values.
+
+    Examples:
+        >>> import pprint
+        >>> d1 = {'a': 1.7, 'b': 2.0, 'c': 1}
+        >>> d2 = {'a': 1.1, 'b': 2.2, 'v': 1}
+        >>> d3 = {'a': 1.1, 'v': 2.3}
+        >>> dflt_func = min
+        >>> agg_funcs = {'a': np.mean, 'v': max}
```
Comment (williamFalcon) [inline pytorch_lightning/loggers/base.py:338]: won't numpy functions slow things down because everything needs to go to cpu? we don't want to move things to CPU for the user ever haha. Every cpu calls slows training down a ton

Comment (Borda) [reply pytorch_lightning/loggers/base.py:338]: so to have it rather as Tensor...

Comment (Borda) [reply pytorch_lightning/loggers/base.py:338]: @alexeykarnachev we may use the Running accum or make another for full accum and extending every N steps and copy existing...
https://discuss.pytorch.org/t/dynamically-extend-the-tensor/39553/2

Comment (Borda) [reply pytorch_lightning/loggers/base.py:338]: so the structure will change from list of dict to dict of tensors but not sure if it makes much faster... also it will allow us to use agg implemented for Torch

Comment (alexeykarnachev) [reply pytorch_lightning/loggers/base.py:338]: Hmmm. I thought, that at this point all values are already on cpu.
Here we call the log procedure:
https://github.com/PyTorchLightning/pytorch-lightning/blob/f7622ebfca45abe7d8d34f2ee2070d6856e24646/pytorch_lightning/trainer/logging.py#L74

And few lines above this call, we transform the metrics ti the scalars:
https://github.com/PyTorchLightning/pytorch-lightning/blob/f7622ebfca45abe7d8d34f2ee2070d6856e24646/pytorch_lightning/trainer/logging.py#L64

So, do we really need tensors here? Metrics which come to the `LightningLoggerBase` are already `item`ed

Comment (alexeykarnachev) [reply pytorch_lightning/loggers/base.py:338]: @Borda , the `TrainerLoggingMixin` forces all metrics to be a scalars. And it seems not very convenient to transform them back to tensors. Besides, as I see, metrics could never be a tensor. For example, I can pass scalar value metric even on the `training_step`, like so:
```python
    def training_step(self, batch, batch_idx):
        loss, logits, _ = self.forward(batch)
        lr = self.trainer.optimizers[0].param_groups[0]['lr']
        log = {'Loss/train': loss, 'Learning-Rate': lr}
        return {'loss': loss, 'log': log}
```
Here, the `lr` is a scalar, it's on cpu and it is not a tensor.

What do you think on this?

Comment (Borda) [reply pytorch_lightning/loggers/base.py:338]: agree, let's get this done and think about speedup later... :]

Comment (williamFalcon) [reply pytorch_lightning/loggers/base.py:338]: ok perfect.

this might actually be the main cause of the minimal speed discrepancy between lightning and pure pytorch.
